### PR TITLE
[Fix] Fix the cast_tensor_type

### PR DESCRIPTION
**Pull request by zhouzaida**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix the `cast_tensor_type` when the type of inputs is not the same as src_type.

## Modification

- Fix the cast_tensor_type in the `mmcv/runner/fp16_utils.py`
- Add two unit tests in the `tests/test_runner/test_fp16.py`

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by teamwong111 — APPROVED** [review]
LGTM

File: mmcv/runner/fp16_utils.py [hunk 1]
```diff
@@ -35,7 +35,9 @@ def cast_tensor_type(inputs, src_type, dst_type):
     if isinstance(inputs, nn.Module):
         return inputs
     elif isinstance(inputs, torch.Tensor):
-        return inputs.to(dst_type)
+        # we need to confirm the type of inputs to be casted is the same as the
```
Comment (ZwwWayne) [inline mmcv/runner/fp16_utils.py:38]: Also update the docstring of the function to describe its behavior in detail.

Comment (zhouzaida) [reply mmcv/runner/fp16_utils.py:47]: done

File: mmcv/runner/fp16_utils.py [hunk 2]
```diff
@@ -24,6 +24,15 @@
 def cast_tensor_type(inputs, src_type, dst_type):
     """Recursively convert Tensor in inputs from src_type to dst_type.
 
+    Note:
+        In v1.4.4 and later, ``cast_tersor_type`` will only convert the
+        torch.Tensor which is consistent with ``src_type`` to the ``dis_type``.
```
Comment (ZwwWayne) [inline mmcv/runner/fp16_utils.py:29]: dis_type -> dst_type

File: mmcv/runner/fp16_utils.py [hunk 3]
```diff
@@ -24,6 +24,15 @@
...
+        Before v1.4.4, it ignores the ``src_type`` argument, leading to some
+        potential problems. For example, invoking
```
Comment (ZwwWayne) [inline mmcv/runner/fp16_utils.py:31]: For example,  ``cast_tensor_type(inputs, torch.float, torch.half)`` will convert all tensors in inputs to ``torch.half`` including those originally in ``torch.Int`` or other types, which is not expected.

File: mmcv/runner/fp16_utils.py [hunk 4]
```diff
@@ -35,7 +44,9 @@ def cast_tensor_type(inputs, src_type, dst_type):
     if isinstance(inputs, nn.Module):
         return inputs
     elif isinstance(inputs, torch.Tensor):
-        return inputs.to(dst_type)
+        # we need to confirm the type of inputs to be casted is the same as the
```
Comment (ZwwWayne) [inline mmcv/runner/fp16_utils.py:47]: we need to ensure that the type of inputs to be casted are the same as the...

**Pull request merged**